src/liboslexec/wide/wide_optexture.cpp

```diff
@@ -561,6 +553,165 @@ dispatch_environment(BatchedRendererServices* bsr, ustring filename,
 }  // namespace
 
 
+static OSL_NOINLINE void
+transformWideTextureGradients(BatchedTextureOutputs& outputs,
+                              Wide<const float> dsdx, Wide<const float> dtdx,
+                              Wide<const float> dsdy, Wide<const float> dtdy)
+{
+    MaskedData resultRef = outputs.result();
+    bool has_derivs      = resultRef.has_derivs();
+    // Matching scalar code path behavior to check only result derivs.
+    ASSERT(has_derivs);
+
+    if (Masked<float>::is(resultRef)) {
+        OSL_FORCEINLINE_BLOCK
+        {
+            MaskedDx<float> drds(resultRef);
+            MaskedDy<float> drdt(resultRef);
+
+            OSL_OMP_PRAGMA(omp simd simdlen(__OSL_WIDTH))
+            for (int i = 0; i < __OSL_WIDTH; ++i) {
+                float drdsVal = drds[i];
+                float drdtVal = drdt[i];
+                float drdx    = drdsVal * dsdx[i] + drdtVal * dtdx[i];
+                float drdy    = drdsVal * dsdy[i] + drdtVal * dtdy[i];
+                drds[i]       = drdx;
+                drdt[i]       = drdy;
+            }
+        }
+    } else {
+        // keep assert out of inlined code
+        ASSERT(Masked<Color3>::is(resultRef));
+        OSL_FORCEINLINE_BLOCK
+        {
+            MaskedDx<Color3> widedrds(resultRef);
+            MaskedDy<Color3> widedrdt(resultRef);
+            OSL_OMP_PRAGMA(omp simd simdlen(__OSL_WIDTH))
+            for (int i = 0; i < __OSL_WIDTH; ++i) {
+                Color3 drdsColor = widedrds[i];
+                Color3 drdtColor = widedrdt[i];
+
+                widedrds[i] = drdsColor * dsdx[i] + drdtColor * dtdx[i];
+                widedrdt[i] = drdsColor * dsdy[i] + drdtColor * dtdy[i];
+            }
+        }
+    }
+
+    MaskedData alphaRef = outputs.alpha();
+    OSL_FORCEINLINE_BLOCK
+    if (alphaRef.valid() && alphaRef.has_derivs()) {
+        MaskedDx<float> dads(alphaRef);
+        MaskedDy<float> dadt(alphaRef);
+        OSL_OMP_PRAGMA(omp simd simdlen(__OSL_WIDTH))
+        for (int i = 0; i < __OSL_WIDTH; ++i) {
+            float dadsVal = dads[i];
+            float dadtVal = dadt[i];
+            float dadx    = dadsVal * dsdx[i] + dadtVal * dtdx[i];
+            float dady    = dadsVal * dsdy[i] + dadtVal * dtdy[i];
+            dads[i]       = dadx;
+            dadt[i]       = dady;
+        }
+    }
+}
+
+static OSL_NOINLINE void
+transformWideTextureGradientsTexture3d(BatchedTextureOutputs& outputs,
+                                       Wide<const Vec3> Pdx,
+                                       Wide<const Vec3> Pdy,
+                                       Wide<const Vec3> Pdz)
+{
+    MaskedData resultRef = outputs.result();
+    bool has_derivs      = resultRef.has_derivs();
+    // Matching scalar code path behavior to check only result derivs.
+    ASSERT(has_derivs);
```

Comment:
I think this should be `OSL_ASSERT` if you want to check all the time, `OSL_DASSERT` if you only want to check for debug builds. I believe the `ASSERT` is just an old alias.

Reply:
I went with OSL_DASSERT for the asserts I left -- in production if the types don't match because we start returning new types, those types will get mis-scaled derivatives, but shouldn't crash, and so I think the debug assert for the developer to notice is appropriate.